nginx/modsecurity/ngx_http_modsecurity.c

```diff
@@ -975,267 +991,170 @@ ngx_http_modsecurity_init_process(ngx_cycle_t *cycle)
-    ngx_table_elt_t                 *h;
-
-    cf = ngx_http_get_module_loc_conf(r, ngx_http_modsecurity);
-    ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
-
-    /* already processed, checking redirect action. */
-    if (ctx && ctx->complete 
-               && r->err_status >= NGX_HTTP_MOVED_PERMANENTLY
-               && r->err_status < 308) {
-
-        /* 3XX load redirect location header so that we can do redirect in phase 3,4 */
-        location = apr_table_get(ctx->req->headers_out, "Location");
-
-        if (location == NULL) {
-            return NGX_HTTP_INTERNAL_SERVER_ERROR;
-        }
+ngx_http_modsecurity_task_offload(ngx_http_request_t *r)
+{
+    ngx_http_modsecurity_thread_ctx_t *thread_ctx;
+    ngx_thread_task_t *task;
 
-        h = ngx_list_push(&r->headers_out.headers);
-        if (h == NULL) {
-            return NGX_HTTP_INTERNAL_SERVER_ERROR;
-        }
+    task = ngx_thread_task_alloc(r->pool, sizeof(ngx_http_modsecurity_thread_ctx_t));
+    if (task == NULL) {
+        return NGX_ERROR;
+    }
 
-        h->hash = 1;
-        h->key.data = (u_char *)"Location";
-        h->key.len = ngx_strlen("Location");
-        h->value.data = (u_char *)location;
-        h->value.len = ngx_strlen(location);
+    thread_ctx = task->ctx;
+    thread_ctx->r = r;
 
-        return ngx_http_next_header_filter(r);
-    }
+    task->handler = ngx_http_modsecurity_thread_func;
+    task->event.handler = ngx_http_modsecurity_thread_completion;
+    task->event.data = r;
 
-    if (r != r->main || !cf->enable || ctx == NULL ||ctx->complete) {
-        return ngx_http_next_header_filter(r);
+    ngx_thread_pool_t* thread_pool = ngx_thread_pool_get((ngx_cycle_t *)ngx_cycle, &thread_pool_name);
+    if (ngx_thread_task_post(thread_pool, task) != NGX_OK) {
+        return NGX_ERROR;
     }
 
-    ngx_log_debug0(NGX_LOG_DEBUG_HTTP, r->connection->log, 0, "modSecurity: header filter");
+    r->main->blocked++;
+    r->aio = 1;
 
-    r->filter_need_in_memory = 1;
     return NGX_OK;
 }
 
 
+static void
+ngx_http_modsecurity_body_handler(ngx_http_request_t *r)
+{
+    r->main->count--;
+    ngx_http_core_run_phases(r);
+}
+
+
+/*
+** [ENTRY POINT] does : this function called by nginx from the request handler
+*/
 static ngx_int_t
-ngx_http_modsecurity_body_filter(ngx_http_request_t *r, ngx_chain_t *in)
+ngx_http_modsecurity_handler(ngx_http_request_t *r)
 {
     ngx_http_modsecurity_loc_conf_t *cf;
     ngx_http_modsecurity_ctx_t      *ctx;
     ngx_int_t                        rc;
-    apr_off_t                        content_length;
-    ngx_chain_t                     *cl, *out;
-    ngx_int_t                        last_buf = 0;
 
     cf = ngx_http_get_module_loc_conf(r, ngx_http_modsecurity);
-    ctx = ngx_http_get_module_ctx(r, ngx_http_modsecurity);
 
-    if (r != r->main || !cf->enable || ctx == NULL || ctx->complete) {
-        return ngx_http_next_body_filter(r, in);
+    /* Process only main request */
```

Comment:
What do you mean by main request?

Reply:
I didn't actually write this comment. It was there before I started. I think they mean don't apply modsec to subrequests, which I believe are requests from the load balancer to the backend. Or else it's some Nginx feature we are not using.